xds/src/main/java/io/grpc/xds/internal/certprovider/ZatarCertificateProvider.java

```diff
@@ -0,0 +1,189 @@
+import java.util.logging.Logger;
+
+/** Implementation of {@link CertificateProvider} for Zatar cert provider. */
+final class ZatarCertificateProvider extends CertificateProvider {
+  private static final Logger logger = Logger.getLogger(ZatarCertificateProvider.class.getName());
+
+  /**
+   * After the previous cert has expired, if we are unable to get new certificates we will report
+   * errors. We will start doing this a few seconds before the previous cert expiry whose value is
+   * given by this constant.
+   */
+  @VisibleForTesting static final long GRACE_INTERVAL_IN_SECONDS = 4L;
+
+  @VisibleForTesting
+  ZatarCertificateProvider(
+      DistributorWatcher watcher,
+      boolean notifyCertUpdates,
+      String directory,
+      String certFile,
+      String privateKeyFile,
+      String trustFile,
+      long refreshIntervalInSeconds,
+      ScheduledExecutorService scheduledExecutorService,
+      TimeProvider timeProvider) {
+    super(watcher, notifyCertUpdates);
+    this.scheduledExecutorService =
+        checkNotNull(scheduledExecutorService, "scheduledExecutorService");
+    this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    this.directory = Paths.get(checkNotNull(directory, "diretory"));
+    this.certFile = checkNotNull(certFile, "certFile");
+    this.privateKeyFile = checkNotNull(privateKeyFile, "privateKeyFile");
+    this.trustFile = checkNotNull(trustFile, "trustFile");
+    this.refreshIntervalInSeconds = refreshIntervalInSeconds;
+    this.syncContext = createSynchronizationContext(directory);
+  }
+
+  private SynchronizationContext createSynchronizationContext(String details) {
+    final InternalLogId logId = InternalLogId.allocate("ZatarCertificateProvider", details);
+    return new SynchronizationContext(
+        new Thread.UncaughtExceptionHandler() {
+          private boolean panicMode;
+
+          @Override
+          public void uncaughtException(Thread t, Throwable e) {
+            logger.log(
+                Level.SEVERE,
+                "[" + logId + "] Uncaught exception in the SynchronizationContext. Panic!",
+                e);
+            panic(e);
+          }
+
+          void panic(final Throwable t) {
+            if (panicMode) {
+              // Preserve the first panic information
+              return;
+            }
+            panicMode = true;
+            close();
+          }
+        });
+  }
+
+  @Override
+  public void start() {
+    scheduleNextRefreshCertificate(/* delayInSeconds= */0);
+  }
+
+  @Override
+  public void close() {
+    if (scheduledHandle != null) {
+      scheduledHandle.cancel();
+      scheduledHandle = null;
+    }
+    getWatcher().close();
+  }
+
+  private void scheduleNextRefreshCertificate(long delayInSeconds) {
+    if (scheduledHandle != null && scheduledHandle.isPending()) {
+      logger.log(Level.SEVERE, "Pending task found: inconsistent state in scheduledHandle!");
+      scheduledHandle.cancel();
+    }
```

Comment:
if `checkAndReloadCertificates` was called from somewhere else and not thru the scheduler then this could happen - but I agree this check is unnecessary if there is no coding bug. Removed